infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/RestoreRedundancyMergeSegments.java

```diff
@@ -1,38 +1,217 @@
+                if (transferredSegments.isEmpty()) {
+                    log.info("State transfer on: {}: No transfer occurred, " +
+                            "try merging the segments.", currentNode);
+                    layoutManagementView.mergeSegments(currentLayout);
+                }
+                // State transfer happened.
+                else {
+                    log.info("State transfer on {}: Transferred segments: {}.", currentNode,
+                            transferredSegments);
+                    // Create a new layout after the segments were transferred.
+                    // After this action is performed a current node will be present
+                    // in all the segments that previously had a status 'TRANSFERRED'.
+                    Layout newLayout = redundancyCalculator.updateLayoutAfterRedundancyRestoration(
+                            transferredSegments, currentLayout);
+
+                    log.info("State transfer on {}: New layout: {}.", currentNode, newLayout);
+
+                    // Merge the segments of the new layout if possible.
+                    if (RedundancyCalculator.canMergeSegments(newLayout)) {
+                        layoutManagementView.mergeSegments(newLayout);
+                    }
+                    // If the segments can't be merged, just propose a new layout.
+                    else {
+                        // Since we seal with a new epoch,
+                        // we also need to bump the epoch of the new layout.
+                        LayoutBuilder builder = new LayoutBuilder(newLayout);
+                        newLayout = builder.setEpoch(currentLayout.getEpoch() + 1).build();
+                        layoutManagementView
+                                .runLayoutReconfiguration(currentLayout, newLayout,
+                                        false);
+                    }
+                }
+                // Return the latest layout.
+                runtime.invalidateLayout();
+                return runtime.getLayoutView().getLayout();
+
+            } catch (WrongEpochException | QuorumUnreachableException | OutrankedException e) {
+                log.warn("Got: {}. Retrying: {} times.", e.getMessage(), retries.get());
+                if (retries.decrementAndGet() < 0) {
+                    throw new RetryExhaustedException("Retries exhausted.", e);
+                } else {
+                    throw new RetryNeededException();
+                }
+            } catch (TransferSegmentException e) {
+                throw new RetryExhaustedException("Transfer segment exception occurred.", e);
+            }
+
+        }).setOptions(retrySettings).run();
+
+    }
+
+    /**
+     * Sets the trim mark on this endpoint's log unit, performs a prefix trim and then compaction.
+     *
+     * @param runtime A current runtime.
+     * @return A retrieved trim mark.
+     */
+    long trimLog(CorfuRuntime runtime) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651453)